llvm/lib/Target/EVM/EVMConstantUnfolding.cpp

```diff
@@ -24,209 +24,243 @@
-      BuildMI(*MI.getParent(), MI, MI.getDebugLoc(), TII->get(Opcode));
-    }
+private:
+  LLVMContext *Context{};
+  const TargetInstrInfo *TII{};
+
+  // The 'second' field can be set to 0 or 1, indicating whether to
+  // optimize for performance or size.
+  using TransformationKey = std::pair<APInt, int>;
+  DenseMap<TransformationKey, std::unique_ptr<TransformationBuilder>>
+      TransformationCache;
+  unsigned OverallCodeReductionSize{0};
+
+  void reduceCodeSizeOn(unsigned Size) {
+    OverallCodeReductionSize += Size;
+    ++NumOfUnfoldings;
+  }
+};
 
-    void buildImm(MachineInstr &MI, const APInt &Val) {
-      unsigned PushOpc = EVM::getPUSHOpcode(Val);
-      auto NewMI = BuildMI(*MI.getParent(), MI, MI.getDebugLoc(),
-                           TII->get(EVM::getStackOpcode(PushOpc)));
-      if (PushOpc != EVM::PUSH0)
-        NewMI.addCImm(ConstantInt::get(Context, Val));
-    }
-  };
+class EVMConstantUnfolding final : public ModulePass {
+public:
+  static char ID;
 
-  MachineFunction *MF = nullptr;
-  const TargetInstrInfo *TII = nullptr;
-  DenseMap<APInt, std::unique_ptr<TransformationBuilder>> TransformationCache;
+  EVMConstantUnfolding() : ModulePass(ID) {
+    initializeEVMConstantUnfoldingPass(*PassRegistry::getPassRegistry());
+  }
 
-  StringRef getPassName() const override { return "EVM constant unfolding"; }
+private:
+  StringRef getPassName() const override { return PASS_NAME; }
 
-  bool runOnMachineFunction(MachineFunction &MF) override;
+  bool runOnModule(Module &M) override;
 
   void getAnalysisUsage(AnalysisUsage &AU) const override {
-    AU.addRequired<MachineLoopInfoWrapperPass>();
-    MachineFunctionPass::getAnalysisUsage(AU);
+    AU.addRequired<MachineModuleInfoWrapperPass>();
+    AU.addPreserved<MachineModuleInfoWrapperPass>();
+    AU.setPreservesAll();
+    ModulePass::getAnalysisUsage(AU);
   }
-
-  const TransformationBuilder *findOptimalTransfomtation(const APInt &Imm);
-
-  bool tryUnfoldConstant(MachineInstr &MI);
 };
 } // end anonymous namespace
 
 char EVMConstantUnfolding::ID = 0;
 
-INITIALIZE_PASS_BEGIN(EVMConstantUnfolding, DEBUG_TYPE, "Constant unfolding",
-                      false, false)
-INITIALIZE_PASS_DEPENDENCY(MachineLoopInfoWrapperPass)
-INITIALIZE_PASS_END(EVMConstantUnfolding, DEBUG_TYPE, "Constant unfolding",
-                    false, false)
+INITIALIZE_PASS_BEGIN(EVMConstantUnfolding, DEBUG_TYPE, PASS_NAME, false, false)
+INITIALIZE_PASS_DEPENDENCY(MachineModuleInfoWrapperPass)
+INITIALIZE_PASS_END(EVMConstantUnfolding, DEBUG_TYPE, PASS_NAME, false, false)
 
-FunctionPass *llvm::createEVMConstantUnfolding() {
+ModulePass *llvm::createEVMConstantUnfolding() {
   return new EVMConstantUnfolding();
 }
 
-const EVMConstantUnfolding::TransformationBuilder *
-EVMConstantUnfolding::findOptimalTransfomtation(const APInt &Imm) {
-  if (auto It = TransformationCache.find(Imm);
+static bool isBetterCandidate(const TransformationBuilder &A,
```

Comment:
I'm thinking - could this be a `operator<` while renaming TransformationBuilder to TransformationCandidate? It will simplify minelement call.